### PR TITLE
added catch to fetching gauges URL

### PR DESCRIPTION
**Pull request by adartulloch**
This PR will handle invalid gauges. If a fetch request returns 500, the gauge will be now be handled.

### CODE REVIEW

**Comment by adartulloch** [comment]
Hmm. So if the gauge is invalid we should be replacing the file image with plaintext rather than any image at all? Am I understanding correctly?

**Comment by adartulloch** [comment]
Also, the new pipe imagery looks so cool! I'm assuming they haven't been merged into master as of yet?

**Comment by Sammidysam** [comment]
First comment: Yes, so on the right instead of the ugly image missing icon, it would say "No Gauge Available".  If the gauges were created by an `img` tag, you could probably set this up as a mix of alt text https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_image_test and JavaScript error hooks https://stackoverflow.com/questions/22051573/how-to-hide-image-broken-icon-using-only-css-html
But a gauge seems to be an `image` element, which I am super confused as to what that actually is.  I could not find much researching just now, so I am not sure what ideas I have really.  Overall, just see if you can change it in the browser then extend it to inside code.
It turns out it is an `svg` element, https://developer.mozilla.org/en-US/docs/Web/SVG/Element/image
And it is an element within `svg-wrap`, which is one giant `svg`.  Maybe this is enough to get a start?  This comment was sprawling because I looked up so much to get to this point.  Let me know if you'd like me to look further.

**Comment by Sammidysam** [comment]
Maybe a good solution is to use an `iframe`, but not sure how well it would work to try to have an `iframe` within an SVG.

**Comment by Sammidysam** [comment]
Second comment: pipes existing or not is a database thing.  I am a little concerned that I do not see the pipes on production - it makes me worry someone messed up the production database.  @rehannan are there pipes missing from production?

**Comment by adartulloch** [comment]
I found similar results about the `image` tag. It's just very much inconvenient that everything is wrapped under this giant SVG because that means I can't render a `div` just to display text. Maybe we'd have to use the `href` to link to another "invalid gauge" image or something. I'm not sure how to go about this :o

**Comment by Sammidysam** [comment]
Just to get it on the paper trail: you told me on Slack that the way to do this is to have a route on this server that returns the "invalid gauge" HTML.  Then you would conditionally set that `href` accordingly.

**Comment by Sammidysam** [comment]
@adartulloch Were you working on this recently and a commit didn't go through?  I remember you talking about it.

**Comment by Sammidysam** [comment]
Thanks for letting me know that gauges are down again.  The console error is fine, but you're saying that your invalid gauge image is indeed shown?  Just confirming.

**Comment by adartulloch** [comment]
Yep, it is!

**Comment by Sammidysam** [comment]
Adar, I'm not sure that the gauges are down - everything looks up to me.  Could you make sure --
Ah yes, your message came in in the meantime.  This is a Data Hub thing I realize - we do not set Access Control Allow Origin to * for the frontend.  I'll quickly make this change and we can see if it helps.  Just a second.

**Comment by Sammidysam** [comment]
@adartulloch It looks like I just added the necessary header.  Feel free to give it a shot again.

File: cwd/app.js [hunk 1]
```diff
@@ -234,6 +234,8 @@
     const flowables = Array.from(document.getElementsByClassName('flowable'));
     const electrons = Array.from(document.getElementsByClassName('electron'));
 
+    //Water droplets need to be set to the right value in the database
```
Comment (Sammidysam) [inline cwd/app.js:246]: Doesn't seem like this is meant to be in this branch?

File: cwd/app.js [hunk 2]
```diff
@@ -269,16 +271,18 @@
...
-            .then(r => r.json())
+            .then(r => {
+              r.json();
+            })
             .then(j => {
               g.data = j;
-
               // Then make call to store the new data into the database.
-              fetch(`http://${API_URL}/glyphs/${obj._id}/gauges/${index + 1}/cache`, {
+              /*fetch(`http://${API_URL}/glyphs/${obj._id}/gauges/${index + 1}/cache`, {
```
Comment (Sammidysam) [inline cwd/app.js]: Why did you comment out the cache?

File: cwd/app.js [hunk 3]
```diff
@@ -269,16 +271,18 @@
...
+      let resStatus = 0;
       if (obj.view && obj.view.gauges) {
         obj.view.gauges.forEach((g, index) => {
-          if (g.data_url) {
+           if (g.data_url) {
             fetch(g.data_url)
-            .then(r => r.json())
+            .then(r => {
+              r.json();
+            })
```
Comment (Sammidysam) [inline cwd/app.js:290]: I *think* this needs to return something or else the later code will break.  Not sure.

File: cwd/server.js [hunk 4]
```diff
@@ -35,5 +35,9 @@ router.use(express.static(__dirname));
 
 router.get('/editor', (req, res) => {
   console.log('Received GET request at /editor');
-  res.sendStatus(200);
+  res.send("okay");
```
Comment (Sammidysam) [inline cwd/server.js:38]: Any particular reason for doing this instead of 200?

File: cwd/server.js [hunk 5]
```diff
@@ -35,5 +35,9 @@ router.use(express.static(__dirname));
...
+});
+
+router.get('/errorgauge', (req, res) => {
```
Comment (Sammidysam) [inline cwd/server.js:41]: I wonder if it may be more efficient to just link to the `/images` path directly later on?

File: cwd/app.js (hunk 1 above)
Comment (adartulloch) [reply cwd/app.js:246]: oops, slipped that in while on the wrong branch

File: cwd/views/index.pug [hunk 6]
```diff
@@ -37,4 +37,5 @@ html
     script const activeGlyphs = !{activeGlyphsStr};
     script const API_URL = "!{api_url}";
     script(src="./dist/cwd-bundle.js")
+    script(src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js")
```
Comment (Sammidysam) [inline cwd/views/index.pug:40]: One thought now that it's working is do you need Ajax, or could you just use the fetch API like I use elsewhere?

File: cwd/app.js (hunk 2 above)
Comment (adartulloch) [reply cwd/app.js]: Outdated, should be fixed in the most recent commit

Comment (Sammidysam) [reply cwd/app.js]: Never mind

File: cwd/views/index.pug (hunk 6 above)
Comment (adartulloch) [reply cwd/views/index.pug:40]: I think I can, not sure of the syntax but I can figure it out. Would you prefer that?

File: cwd/server.js (hunk 5 above)
Comment (adartulloch) [reply cwd/server.js:41]: Yep meant to eliminate all this

File: cwd/server.js (hunk 4 above)
Comment (adartulloch) [reply cwd/server.js:38]: Fixed

File: cwd/views/index.pug (hunk 6 above)
Comment (Sammidysam) [reply cwd/views/index.pug:40]: Yeah, I like for either all requests to be AJAX or all requests to be fetch api.  I like the fetch api better honestly.

Comment (adartulloch) [reply cwd/views/index.pug:40]: Sounds good! I'll clean this all up.

File: cwd/app.js [hunk 7]
```diff
@@ -269,16 +278,16 @@
...
-          if (g.data_url) {
+           if (g.data_url) {
             fetch(g.data_url)
             .then(r => r.json())
             .then(j => {
               g.data = j;
-
               // Then make call to store the new data into the database.
-              fetch(`${API_URL}/glyphs/${obj._id}/gauges/${index + 1}/cache`, {
+              fetch(`http://${API_URL}/glyphs/${obj._id}/gauges/${index + 1}/cache`, {
```
Comment (Sammidysam) [inline cwd/app.js:290]: This must be removed, or else it breaks when we use HTTPS in production.

File: cwd/app.js [hunk 8]
```diff
@@ -38,14 +38,14 @@
   const SHOW_ONE_TITLE = 1;
 
   /** The amount of time in seconds between views in kiosk mode. */
-  const VIEW_DURATION = 10;
+  const VIEW_DURATION = 2;
```
Comment (Sammidysam) [inline cwd/app.js:41]: Please revert this

File: cwd/app.js [hunk 9]
```diff
@@ -38,14 +38,14 @@
...
 
   /** Default message to be shown only if the DB has no messages. */
   const DEFAULT_MESSAGE = "Welcome to Citywide Dashboard!";
 
   /**
    * Example of accessing API:
-   * fetch(`${API_URL}/glyphs`)
+   * fetch(`http://${API_URL}/glyphs`)
```
Comment (Sammidysam) [inline cwd/app.js:48]: Please revert this - we cannot use this prefix on production

**Pull request merged**